backfill.sh

```diff
@@ -0,0 +1,140 @@
+runpath=$(realpath $runpath) # realpath needs to be installed :(
+mkdir -p "$runpath"
+
+shift # download-api-raml.sh takes an arg :P
+
+# confirm
+
+echo "backfill.sh
+
+this script will:
+1. pull latest article-xml from elifesciences/elife-article-xml (large repo)
+2. download any missing/unpublished articles after consulting Lax (needs /srv/lax, s3 auth)
+3. create a 'backfill-run' directory with symbolic links to the xml to be processed
+4. generate article-json from ALL xml in the ./articles-xml/articles/ directory (long process)
+5. validate all generated article-json, failing if any are invalid
+6. force an INGEST into Lax for all valid articles (needs /srv/lax)"
+
+read -p "any key to continue (ctrl-c to quit) "
+
+# begin
+
+. download-elife-xml.sh
+
+# activate venv
+set +o nounset; . install.sh; set -o nounset; # virtualenv has unset vars we can't control
+
+# where to find xml on the fs
+xmlrepodir="$srcdir/article-xml/articles"
+
+# where to download unpublished xml to
+unpubxmldir="$srcdir/unpub-article-xml"
+mkdir -p "$unpubxmldir" # (create if necessary)
+
+# where generated article-json will be stored
+ajsondir="$runpath/ajson"
+mkdir -p "$ajsondir"
+
+# where the results of validation will be stored
+validdir="$ajsondir/valid"
+invaliddir="$ajsondir/invalid"
+patcheddir="$ajsondir/patched"
+
+# because we can choose an existing directory for the run
+# ensure the results of any previous run are empty
+rm -rf "$validdir" "$invaliddir" "$patcheddir"
+mkdir "$validdir" "$invaliddir" "$patcheddir"
+
+
+#
+#
+#
+
+# switch to the run dir
+cd "$runpath"
+
+# iterate over response from lax about the articles it knows about
+# (should be *all* articles, lax is now the publishing authority)
+# https://www.cyberciti.biz/faq/unix-linux-bash-read-comma-separated-cvsfile/
+OLDIFS=$IFS
+IFS=,
+/srv/lax/manage.sh --skip-install report all-article-versions-as-csv | while read msid version remotepath
+do
+    # ll: elife-00003-v1.xml
+    fname="elife-$msid-v$version.xml" 
+
+    # ll: /home/user/bot-lax/article-xml/articles/elife-00003-v1.xml
+    xmlpath="$xmlrepodir/$fname"
+    
+    # ll: /home/user/bot-lax/unpub-article-xml/elife-00003-v1.xml
+    xmlunpubpath="$unpubxmldir/$fname"
+    
+    # we look in both places for xml and if it's in neither, we download it
+
+    if [ ! -f $xmlpath ] && [ ! -f $xmlunpubpath ]; then
+        # lax doesn't know where the remote location and it's not on the fs
+        if [ $remotepath = "no-location-stored" ]; then
+            errcho "$fname not found, skipping"
+            continue
+        fi
+
+        # xml absent, download it
+        # download.py reuses code in the adaptor and does an authenticated requests to s3
+        python $srcdir/src/download.py "$remotepath" "$xmlunpubpath"
+    fi
+
+    #errcho "linking $fname"
+
+    # link it in to the run dir
+    if [ -f $xmlpath ]; then
+        ln -sfT $xmlpath $fname
+    else
+        ln -sfT $xmlunpubpath $fname
+    fi
```

Comment:
ok, so if the `article-xml` repository has an XML we take that one, otherwise it means it's either a) unpublished or b) not yet in the repo, in both cases downloading it from the remote path is the best choice.

This doesn't cater for silent corrections however, if the sha of `article-xml` does not have them (because it's slightly out of date) the old version in there will be used instead of the one silently corrected. This has to be solved as it is dangerous. Perhaps we should switch to the `master` of `article-xml` for the backfill? We won't have the guarantee that all articles are definitely valid with this code as the more recent ones haven't been tested, but they got published so they were valid before - it's a small price to pay.